File: scripts/hooks/session_init.py
#!/usr/bin/env python3
"""VaultOps Brain — SessionStart hook.

Fires on session startup/resume/clear/compact.
Initializes session state and loads any active IN_PROGRESS task from the vault.
"""

import json
import os
import sys
from datetime import datetime, timezone

# Add hooks dir and parent for imports
_hooks_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, _hooks_dir)
sys.path.insert(0, os.path.dirname(_hooks_dir))

from brain_state import is_vaultops_project, load_state, save_state, _DEFAULT_STATE  # noqa: E402
from vaultops_mcp_server import (  # noqa: E402
    _tool_get_context,
    _tool_get_replay,
    _exec_path,
    _read_file,
    _parse_frontmatter,
    EXEC_JOURNAL,
    SPRINTS_DIR,
)


def main() -> None:
    project_path = os.getcwd()

    if not is_vaultops_project(project_path):
        return

    # Initialize fresh state
    state = dict(_DEFAULT_STATE)
    state["session_start"] = datetime.now(timezone.utc).isoformat()

    # Load current vault context to find active tasks
    try:
        result = _tool_get_context({"project_path": project_path})
        content = result.get("content", [{}])
        if content and isinstance(content, list):
            text = content[0].get("text", "")
            data = json.loads(text) if text else {}
        else:
            data = {}
    except Exception:
        data = {}

    active_tasks = data.get("active_tasks", [])
    task_summary = data.get("task_summary", {})

    # If there's an active task, resume it
    lines = ["[VaultOps Brain \u2014 Active]"]

    if active_tasks:
        focus = active_tasks[0]
        state["active_task_id"] = focus["id"]
        state["phase"] = "working"
        lines.append(f"Resuming {focus['id']}: {focus['task']} [{focus.get('priority', '')}]")
        if len(active_tasks) > 1:
            lines.append(f"({len(active_tasks)} tasks in progress)")
    else:
        lines.append("No active task \u2014 will auto-create on first work intent.")

    blocked = data.get("blocked_tasks", [])
    if blocked:
        lines.append(f"\u26a0\ufe0f {len(blocked)} blocked task(s)")

    total = task_summary.get("total", 0)
    done = task_summary.get("done", 0)
    if total:
        lines.append(f"Vault: {done}/{total} tasks done")

    # Surface learning insight if available
    try:
        from vaultops_mcp_server import _resolve_vault_project, _parse_frontmatter, _read_file as _rf
        vault_project = _resolve_vault_project(project_path)
        if vault_project:
            patterns_path = os.path.join(vault_project, "08-Execution", "Learnings", "Patterns.md")
            if os.path.isfile(patterns_path):
                pfm, _ = _parse_frontmatter(_rf(patterns_path))
                conf = pfm.get("confidence", "low")
                analyzed = pfm.get("total_tasks_analyzed", 0)
                if conf in ("medium", "high") and analyzed:
                    lines.append(f"Learning: {conf} confidence ({analyzed} tasks analyzed)")
    except Exception:
        pass

    # ── Auto-Replay: compact digest of recent activity ──────────────
    try:
        replay = _tool_get_replay({"project_path": project_path, "hours": 24})
        replay_content = replay.get("content", [{}])
        if replay_content and isinstance(replay_content, list):
            replay_text = replay_content[0].get("text", "")
            # Extract JSON from receipt (after the visual part)
            for segment in replay_text.split("\n\n"):
                s = segment.strip()
                if s.startswith("{"):
                    try:
                        rd = json.loads(s)
                        parts = []
                        if rd.get("tasks_completed"):
                            parts.append(f"{rd['tasks_completed']} task(s) completed")
                        if rd.get("tasks_created"):
                            parts.append(f"{rd['tasks_created']} created")
                        if rd.get("stale_docs"):
                            parts.append(f"{rd['stale_docs']} stale doc(s)")
                        if parts:
                            lines.append(f"Last 24h: {', '.join(parts)}")
                        break
                    except (json.JSONDecodeError, TypeError):
                        continue
    except Exception:
        pass

    # ── Sprint deadline check ─────────────────────────────────────
    try:
        exec_dir, err = _exec_path(project_path)
        if not err:
            sprints_path = os.path.join(exec_dir, SPRINTS_DIR)
            if os.path.isdir(sprints_path):
                today = datetime.now(timezone.utc).strftime("%Y-%m-%d")
                for fname in os.listdir(sprints_path):
                    if fname.startswith("Sprint-") and fname.endswith(".md") and "Retro" not in fname:
                        sc = _read_file(os.path.join(sprints_path, fname))
                        sfm, _ = _parse_frontmatter(sc)
                        end_date = sfm.get("end_date", "")
                        sprint_num = sfm.get("number", fname.replace("Sprint-", "").replace(".md", ""))
                        if end_date and end_date <= today:
                            # Check if retro exists
                            retro_file = os.path.join(sprints_path, f"Sprint-{sprint_num}-Retro.md")
                            if not os.path.isfile(retro_file):
                                lines.append(f"\U0001f4e2 Sprint {sprint_num} ended ({end_date}) — run /vault:retro {sprint_num}")
    except Exception:
        pass

    lines.append('Auto-tracking: tasks, journal, docs. Say "don\'t track" to suppress.')

    save_state(project_path, state)

    print(json.dumps({"additionalContext": "\n".join(lines)}))


if __name__ == "__main__":
    main()

File: src/hooks/session-init.ts
/**
 * VaultOps Brain — SessionStart hook.
 *
 * Fires on session startup/resume/clear/compact.
 * Initializes session state and loads any active IN_PROGRESS task from the vault.
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { isVaultopsProject, getDefaultState, saveState } from "./shared/brain-state.js";
import { getContext } from "../tools/core.js";
import { execPath, resolveVaultProject } from "../vault/resolve.js";
import { readFileOrNull, fileExists } from "../fs/read.js";
import { parseFrontmatter } from "../fs/frontmatter.js";
import { SPRINTS_DIR } from "../constants.js";
import { getSyncConfig, getLastSyncStatus, hasGitRepo } from "../sync.js";

function main(): void {
  const projectPath = process.cwd();

  if (!isVaultopsProject(projectPath)) return;

  // Initialize fresh state
  const state = getDefaultState();
  state.session_start = new Date().toISOString();

  // Load current vault context to find active tasks
  let data: Record<string, unknown> = {};
  try {
    const result = getContext({ project_path: projectPath });
    const content = result.content as Array<{ text?: string }>;
    if (content?.[0]?.text) {
      data = JSON.parse(content[0].text);
    }
  } catch { /* ignore */ }

  const activeTasks = (data.active_tasks ?? []) as Array<Record<string, string>>;
  const taskSummary = (data.task_summary ?? {}) as Record<string, number>;

  const lines: string[] = [];

  // Active task — primary context
  if (activeTasks.length) {
    const focus = activeTasks[0];
    state.active_task_id = focus.id;
    state.phase = "working";
    const extra = activeTasks.length > 1 ? ` (+${activeTasks.length - 1} more)` : "";
    lines.push(`[VaultOps] ${focus.id}: ${focus.task}${extra}`);
  }

  const total = taskSummary.total ?? 0;
  const done = taskSummary.done ?? 0;

  // Urgent items only (blocked tasks, sprint deadline ≤ 3 days, sync conflicts)
  const urgentParts: string[] = [];

  const blocked = (data.blocked_tasks ?? []) as unknown[];
  if (blocked.length) urgentParts.push(`${blocked.length} blocked`);

  // Sprint deadline — only if within 3 days
  try {
    const [execDir, err] = execPath(projectPath);
    if (!err && execDir) {
      const sprintsPath = path.join(execDir, SPRINTS_DIR);
      if (fs.existsSync(sprintsPath) && fs.statSync(sprintsPath).isDirectory()) {
        const today = new Date().toISOString().slice(0, 10);
        const threeDays = new Date(Date.now() + 3 * 86400000).toISOString().slice(0, 10);
        for (const fname of fs.readdirSync(sprintsPath)) {
          if (fname.startsWith("Sprint-") && fname.endsWith(".md") && !fname.includes("Retro")) {
            const sc = readFileOrNull(path.join(sprintsPath, fname)) ?? "";
            const { frontmatter: sfm } = parseFrontmatter(sc);
            const endDate = String(sfm.end_date ?? "");
            const sprintNum = String(sfm.number ?? fname.replace("Sprint-", "").replace(".md", ""));
            if (endDate && endDate <= threeDays) {
              const retroFile = path.join(execDir, SPRINTS_DIR, `Sprint-${sprintNum}-Retro.md`);
              if (!fileExists(retroFile)) {
                urgentParts.push(`Sprint ${sprintNum} ends ${endDate}`);
              }
            }
          }
        }
      }
    }
  } catch { /* ignore */ }

  // Sync conflicts only (not routine sync status)
  try {
    const syncCfg = getSyncConfig(projectPath);
    const vaultProject = resolveVaultProject(projectPath);
    if (syncCfg.enabled && vaultProject && hasGitRepo(vaultProject)) {
      const status = getLastSyncStatus(projectPath, vaultProject);
      if (status.hasConflicts) {
        urgentParts.push("sync conflicts — see Conflict Report.md");
      }
    }
  } catch { /* ignore */ }

  if (urgentParts.length) {
    lines.push(`[!] ${urgentParts.join(" · ")}`);
  }

  if (!activeTasks.length) {
    if (total) {
      lines.push(`[VaultOps] ${done}/${total} tasks done. Auto-tracking active.`);
    } else {
      lines.push("[VaultOps] Auto-tracking active.");
    }
  }

  saveState(projectPath, state);

  process.stdout.write(JSON.stringify({ additionalContext: lines.join("\n") }) + "\n");
}

main();
